modules/ctl/src/main/scala/free/interp.scala

```diff
@@ -0,0 +1,119 @@
+package gem.ctl
+package free
+
+import scalaz._, Scalaz._
+import scalaz.effect._
+
+import gem.ctl.low.io._
+import gem.ctl.free.ctl._
+
+/** An interpreter for the control language. */
+object interpreter {
+
+  /** Configuration for the interpreter. */
+  trait Config {
+    def verbose: Boolean
+    def server:  Server
+  }
+
+  case class InterpreterState(indentation: Int, machineHostCache: Map[Host.Machine, String]) {
+    def indent  = copy(indentation = indentation + 1)
+    def outdent = copy(indentation = indentation - 1)
+  }
+  object InterpreterState {
+    val initial = InterpreterState(0, Map.empty)
+  }
+
+  /**
+   * Construct an interpreter of `CtlIO` given a `Config` and an `IORef` for the initial state. We
+   * carry our state in an `IORef` because it needs to be visible from multiple threads.
+   */
+  def interpreter(c: Config, state: IORef[InterpreterState]) = λ[CtlOp ~> EitherT[IO, Int, ?]] {
```

Comment:
@jdnavarro I think there is consensus that using Scala just as a more modern Java is out of the question.  That wouldn't improve our current situation where we are afraid to make changes to the codebase because we cannot control all the unintended side-effects.  I think it's more a question of how do we all reach the level of expertise that we need, do we all understand the benefits and justification of doing something one way or another, etc.